src/wallet/walletutil.cpp

```diff
@@ -23,5 +23,11 @@ fs::path GetWalletDir()
         }
     }
 
+    // Ensure that the directory does not end with a trailing separator to avoid
+    // creating two Berkeley environments in the same directory
+    while (fs::detail::is_directory_separator(path.string().back())) {
```

Comment:
I think you could add above:
```cpp
if (gArgs.IsArgSet("-walletdir")) {
    path = gArgs.GetArg("-walletdir", "");
    boost::system::error_code error;
    path = fs::canonical(path, error);
    if (error || !fs::is_directory(path)) {
        ...
```
`fs::canonical` checks if it exists and also cleans the path. For instance, locally calling `canonical("/Users/promag/.///.//")` gives `/Users/promag`.

Reply:
Ah `fs::canonical` is even better! What do you think about me adding this to `WalletInit::Verify` [here](https://github.com/bitcoin/bitcoin/blob/0.17/src/wallet/init.cpp#L180-L189) and then `gArgs.ForceSetArg` with the new clean path? 
That way we keep the error checking in one place and the path gets cleaned up once, rather than every time `GetWalletDir` is called.

Reply:
Sounds good, you could replace `fs::exists` with `fs::canonical` in https://github.com/bitcoin/bitcoin/blob/465a583f9dcc2383eb20bbb98bffa66fe82f2cde/src/wallet/init.cpp#L182

Reply:
The nuance with that approach is that we want to avoid transforming a relative file path to an absolute one, per @ryanofsky's commit message there: 

> Specifying paths relative to the current working directory in a daemon process can be dangerous, because files can fail to be located even if the configuration doesn't change, but the daemon is started up differently.

I'm thinking to first check `wallet_dir.is_absolute()` and then use `fs::canonical`